File: src/java_pecs/Shop.java
package java_pecs;

import java.util.ArrayList;
import java.util.List;

public class Shop {
    public static void main(String[] args) {
        List<Product> productsSrc = new ArrayList<>();
        productsSrc.add(new Camera());
        productsSrc.add(new Phone());
        productsSrc.add(new Camera());

        List<Product> listDst = new ArrayList<>();

        Helper helper = new Helper();
        helper.copy(productsSrc,listDst);

        for(Object p: listDst){
            System.out.println(p.getClass());
        }
    }
}

class Helper{
    public void copy(List<? extends Product> src, List<? super Product> dest){
        for (Product p : src){
            dest.add(p);
        }
    }
}

File: src/kotlin_pecs/Shop.kt
package kotlin_pecs

import java_pecs.Camera
import java_pecs.Phone
import java_pecs.Product
import java.util.*

    fun main(args: Array<String>) {
        val productsSrc: MutableList<Product> = ArrayList()
        productsSrc.add(Camera())
        productsSrc.add(Phone())
        productsSrc.add(Camera())
        val listDst: MutableList<Product> = ArrayList()
        val helper = Helper()
        helper.copy(productsSrc, listDst)
        for (p in listDst) {
            println(p.javaClass)
        }
    }


internal class Helper {
    fun copy(src: List<Product>, dest: MutableList<in Product>) {
        for (p in src) {
            dest.add(p)
        }
    }
}